AssertionError: Can't parse version default/fix/fluidsim-core-0.4.1

I can't install a package (https://github.com/exabl/snek5000) because of a setuptools_scm error (AssertionError: Can't parse version default/fix/fluidsim-core-0.4.1). What can I do?

Note: I cloned the Github package with hg.

```
pierre@gre ~/Dev/snek5000 master $ hg sum                                       
parent: 285:7def824c6333 default/master tip
 Merge pull request #20 from exabl/docs/material
branch: default
bookmarks: *master
lib/Nek5000/run/.gitignore: unsupported ignore pattern '!.gitignore'
lib/Nek5000/3rd_party/.gitignore: unsupported ignore pattern '!.gitignore'
commit: (clean)
update: 13 new changesets, 4 branch heads (merge)
phases: 286 draft

pierre@gre ~/Dev/snek5000 master $ hg book                                      
   enh/fluidsim-core         274:a10bc6086557
   enh/paraview              208:762b250ebd1e
   enh/pytest-cov            236:00c4e8d71fa0
   enh/snake-nek             261:b6d48d986176
   enh/templates             196:f7b720daaa31
   enh/upgrade-v19           213:1a195421e67e
   fix/activate_paths        201:589c961eac38
   fix/ci                    169:eca44820047d
   fix/fluidsim-0.3.3        243:a3d69ea3e94f
   fix/fluidsim-core-0.4.1   281:7bcf2618917b
   gh-pages                  191:5c5fd8465ac0
 * master                    285:7def824c6333

pierre@gre ~/Dev/snek5000 master $ pip install -e .                             
Obtaining file:///home/pierre/Dev/snek5000
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  ERROR: Command errored out with exit status 1:
   command: /home/pierre/.pyenv/versions/3.8.2/bin/python3.8 /home/pierre/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmplbof4igl
       cwd: /home/pierre/Dev/snek5000
  Complete output (50 lines):
  /tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/dist.py:634: UserWarning: Usage of dash-separated 'project-urls' will not be supported in future versions. Please use the underscore name 'project_urls' instead
    warnings.warn(
  /tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/dist.py:634: UserWarning: Usage of dash-separated 'max-line-length' will not be supported in future versions. Please use the underscore name 'max_line_length' instead
    warnings.warn(
  /tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/version.py:88: UserWarning: tag 'default/master' no version found
    warnings.warn("tag {!r} no version found".format(tag))
  /tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/version.py:88: UserWarning: tag 'default/fix/fluidsim-core-0.4.1' no version found
    warnings.warn("tag {!r} no version found".format(tag))
  Traceback (most recent call last):
    File "/home/pierre/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 280, in <module>
      main()
    File "/home/pierre/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 263, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/home/pierre/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 114, in get_requires_for_build_wheel
      return hook(config_settings)
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 149, in get_requires_for_build_wheel
      return self._get_build_requires(
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 130, in _get_build_requires
      self.run_setup()
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 145, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 10, in <module>
      setup()
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
      return distutils.core.setup(**attrs)
    File "/home/pierre/.pyenv/versions/3.8.2/lib/python3.8/distutils/core.py", line 108, in setup
      _setup_distribution = dist = klass(attrs)
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 432, in __init__
      _Distribution.__init__(self, {
    File "/home/pierre/.pyenv/versions/3.8.2/lib/python3.8/distutils/dist.py", line 292, in __init__
      self.finalize_options()
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 735, in finalize_options
      ep(self)
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/integration.py", line 61, in infer_version
      dist.metadata.version = _get_version(config)
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 173, in _get_version
      parsed_version = _do_parse(config)
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 127, in _do_parse
      version = _version_from_entrypoints(config) or _version_from_entrypoints(
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 65, in _version_from_entrypoints
      version = _call_entrypoint_fn(root, config, ep.load())
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 46, in _call_entrypoint_fn
      return fn(root, config=config)
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/hg.py", line 63, in parse
      return _hg_tagdist_normalize_tagcommit(config, tag, dist, node, branch)
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/hg.py", line 28, in _hg_tagdist_normalize_tagcommit
      return meta(
    File "/tmp/pip-build-env-4fhq9xco/overlay/lib/python3.8/site-packages/setuptools_scm/version.py", line 212, in meta
      assert parsed_version is not None, "Can't parse version %s" % tag
  AssertionError: Can't parse version default/fix/fluidsim-core-0.4.1
  ----------------------------------------
WARNING: Discarding file:///home/pierre/Dev/snek5000. Command errored out with exit status 1: /home/pierre/.pyenv/versions/3.8.2/bin/python3.8 /home/pierre/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmplbof4igl Check the logs for full command output.
ERROR: Command errored out with exit status 1: /home/pierre/.pyenv/versions/3.8.2/bin/python3.8 /home/pierre/.pyenv/versions/3.8.2/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmplbof4igl Check the logs for full command output.
```

--- a/.github/workflows/python-tests.yml
+++ b/.github/workflows/python-tests.yml
@@ -40,6 +40,8 @@ jobs:
       - run: pip install -U setuptools
         if: matrix.python_version != 'msys2'
       - run: pip install -e .[toml] pytest
+      - run: $(hg debuginstall --template "{pythonexe}") -m pip install hg-git --user
+        if: matrix.os == "ubuntu-latest"
       - run: pytest
 
   check_selfinstall:

--- a/.github/workflows/python-tests.yml
+++ b/.github/workflows/python-tests.yml
@@ -42,7 +42,7 @@ jobs:
         if: matrix.python_version != 'msys2'
       - run: pip install -e .[toml] pytest
       - run: $(hg debuginstall --template "{pythonexe}") -m pip install hg-git --user
-        if: matrix.os == "ubuntu-latest"
+        if: matrix.os == 'ubuntu-latest'
       - run: pytest
 
   check_selfinstall:

--- a/.github/workflows/python-tests.yml
+++ b/.github/workflows/python-tests.yml
@@ -41,7 +41,11 @@ jobs:
       - run: pip install -U setuptools
         if: matrix.python_version != 'msys2'
       - run: pip install -e .[toml] pytest
-      - run: $(hg debuginstall --template "{pythonexe}") -m pip install hg-git --user
+      # pip2 is needed because Mercurial uses python2 on Ubuntu 20.04
+      - run: |
+          curl https://bootstrap.pypa.io/pip/2.7/get-pip.py --output get-pip.py
+          sudo python2 get-pip.py
+          $(hg debuginstall --template "{pythonexe}") -m pip install hg-git --user
         if: matrix.os == 'ubuntu-latest'
       - run: pytest
 

--- a/src/setuptools_scm/git.py
+++ b/src/setuptools_scm/git.py
@@ -235,6 +235,8 @@ def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
 
     if ret == 0:
         tag, distance, node, dirty = _git_parse_describe(out)
+        if distance == 0 and not dirty:
+            distance = None
     else:
         # If 'git git_describe_command' failed, try to get the information otherwise.
         tag = "0.0"
@@ -244,8 +246,6 @@ def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
         else:
             distance = wd.count_all_nodes()
             node = "g" + node
-        if distance == 0:
-            distance = None
         dirty = wd.is_dirty()
 
     branch = wd.get_branch()

--- a/src/setuptools_scm/git.py
+++ b/src/setuptools_scm/git.py
@@ -19,7 +19,6 @@ class GitWorkdir:
     COMMAND = "git"
 
     def __init__(self, path):
-        require_command(self.COMMAND)
         self.path = path
 
     def do_ex(self, cmd):
@@ -27,6 +26,7 @@ class GitWorkdir:
 
     @classmethod
     def from_potential_worktree(cls, wd):
+        require_command(cls.COMMAND)
         wd = os.path.abspath(wd)
         real_wd, _, ret = do_ex("git rev-parse --show-prefix", wd)
         real_wd = real_wd[:-1]  # remove the trailing pathsep
@@ -54,7 +54,11 @@ class GitWorkdir:
         branch, err, ret = self.do_ex("git rev-parse --abbrev-ref HEAD")
         if ret:
             trace("branch err", branch, err, ret)
-            return
+            # TODO: understand the diff between these 2 commands
+            branch, err, ret = self.do_ex("git symbolic-ref --short HEAD")
+            if ret:
+                trace("branch err (symbolic-ref)", branch, err, ret)
+                branch = None
         return branch
 
     def get_head_date(self):
@@ -76,9 +80,9 @@ class GitWorkdir:
         self.do_ex("git fetch --unshallow")
 
     def node(self):
-        rev_node, _, ret = self.do_ex("git rev-parse --verify --quiet HEAD")
+        node, _, ret = self.do_ex("git rev-parse --verify --quiet HEAD")
         if not ret:
-            return rev_node[:7]
+            return node[:7]
 
     def count_all_nodes(self):
         revs, _, _ = self.do_ex("git rev-list HEAD")
@@ -93,6 +97,7 @@ class GitWorkdirHgClient(GitWorkdir):
 
     @classmethod
     def from_potential_worktree(cls, wd):
+        require_command(cls.COMMAND)
         root, _, ret = do_ex("hg root", wd)
         if ret:
             return
@@ -123,9 +128,9 @@ class GitWorkdirHgClient(GitWorkdir):
         pass
 
     def get_hg_node(self):
-        rev_node, _, ret = self.do_ex("hg log -r . -T {node}")
+        node, _, ret = self.do_ex("hg log -r . -T {node}")
         if not ret:
-            return rev_node
+            return node
 
     def node(self):
         hg_node = self.get_hg_node()
@@ -154,9 +159,9 @@ class GitWorkdirHgClient(GitWorkdir):
         hg_tags, _, ret = self.do_ex(
             "hg log -r reverse(ancestors(.)) -T {tags}{if(tags, ' ', '')}"
         )
-
         if ret:
             return None, None, None
+        hg_tags = hg_tags.split()
 
         git_tags = {}
         with open(os.path.join(self.path, ".hg/git-tags"), "r") as file:
@@ -170,13 +175,13 @@ class GitWorkdirHgClient(GitWorkdir):
             if tag in git_tags:
                 break
 
-        out, _, ret = self.do_ex("hg log -r .:" + tag + " -T .")
+        out, _, ret = self.do_ex("hg log -r .::" + tag + " -T .")
         if ret:
             return None, None, None
         distance = len(out) - 1
 
-        rev_node = self.node()
-        desc = f"{tag}-{distance}-g{rev_node}"
+        node = self.node()
+        desc = f"{tag}-{distance}-g{node}"
 
         if self.is_dirty():
             desc += "-dirty"
@@ -193,7 +198,7 @@ def warn_on_shallow(wd):
 def fetch_on_shallow(wd):
     """experimental, may change at any time"""
     if wd.is_shallow():
-        warnings.warn('"%s" was shallow, git fetch was used to rectify')
+        warnings.warn(f'"{wd.path}" was shallow, git fetch was used to rectify')
         wd.fetch_shallow()
 
 
@@ -224,36 +229,26 @@ def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
         describe_command = config.git_describe_command
 
     if describe_command is not None:
-        out, unused_err, ret = wd.do_ex(describe_command)
+        out, _, ret = wd.do_ex(describe_command)
     else:
-        out, unused_err, ret = wd.default_describe()
+        out, _, ret = wd.default_describe()
 
-    if ret:
+    if ret == 0:
+        tag, distance, node, dirty = _git_parse_describe(out)
+    else:
         # If 'git git_describe_command' failed, try to get the information otherwise.
-        branch, branch_err, branch_ret = wd.do_ex("git symbolic-ref --short HEAD")
-
-        if branch_ret:
-            branch = None
-
-        rev_node = wd.node()
-        dirty = wd.is_dirty()
-
         tag = "0.0"
-
-        if rev_node is None:
+        node = wd.node()
+        if node is None:
             distance = 0
         else:
             distance = wd.count_all_nodes()
-            branch = wd.get_branch()
-            node = "g" + rev_node
-
-    else:
-        tag, distance, node, dirty = _git_parse_describe(out)
-
-        branch = wd.get_branch()
+            node = "g" + node
         if distance == 0:
             distance = None
+        dirty = wd.is_dirty()
 
+    branch = wd.get_branch()
     node_date = wd.get_head_date() or date.today()
 
     return meta(

--- a/src/setuptools_scm/git.py
+++ b/src/setuptools_scm/git.py
@@ -1,28 +1,20 @@
+from datetime import datetime, date
+import os
+from os.path import isfile, join, samefile
+import warnings
+
 from .config import Configuration
 from .utils import do_ex, trace, require_command
 from .version import meta
-from datetime import datetime, date
-import os
-from os.path import isfile, join
-import warnings
-
-
-from os.path import samefile
-
+from .scm_workdir import Workdir
 
 DEFAULT_DESCRIBE = "git describe --dirty --tags --long --match *[0-9]*"
 
 
-class GitWorkdir:
+class GitWorkdir(Workdir):
     """experimental, may change at any time"""
 
     COMMAND = "git"
-
-    def __init__(self, path):
-        self.path = path
-
-    def do_ex(self, cmd):
-        return do_ex(cmd, cwd=self.path)
 
     @classmethod
     def from_potential_worktree(cls, wd):
@@ -92,109 +84,6 @@ class GitWorkdir:
         return self.do_ex(DEFAULT_DESCRIBE)
 
 
-class GitWorkdirHgClient(GitWorkdir):
-    COMMAND = "hg"
-
-    @classmethod
-    def from_potential_worktree(cls, wd):
-        require_command(cls.COMMAND)
-        root, _, ret = do_ex("hg root", wd)
-        if ret:
-            return
-        return cls(root)
-
-    def is_dirty(self):
-        out, _, _ = self.do_ex("hg id -T '{dirty}'")
-        return bool(out)
-
-    def get_branch(self):
-        branch, err, ret = self.do_ex("hg id -T {bookmarks}")
-        if ret:
-            trace("branch err", branch, err, ret)
-            return
-        return branch
-
-    def get_head_date(self):
-        date_part, err, ret = self.do_ex("hg log -r . -T {shortdate(date)}")
-        if ret:
-            trace("head date err", date_part, err, ret)
-            return
-        return datetime.strptime(date_part, r"%Y-%m-%d").date()
-
-    def is_shallow(self):
-        return False
-
-    def fetch_shallow(self):
-        pass
-
-    def get_hg_node(self):
-        node, _, ret = self.do_ex("hg log -r . -T {node}")
-        if not ret:
-            return node
-
-    def node(self):
-        hg_node = self.get_hg_node()
-        if hg_node is None:
-            return
-
-        git_node = None
-        with open(os.path.join(self.path, ".hg/git-mapfile"), "r") as file:
-            for line in file:
-                if hg_node in line:
-                    git_node, hg_node = line.split()
-                    break
-
-        if git_node is None:
-            # trying again after hg -> git
-            self.do_ex("hg gexport")
-            return self.node()
-
-        return git_node[:7]
-
-    def count_all_nodes(self):
-        revs, _, _ = self.do_ex("hg log -r 'ancestors(.)' -T '.'")
-        return len(revs) + 1
-
-    def default_describe(self):
-        """
-        Tentative to reproduce the output of
-
-        `git describe --dirty --tags --long --match *[0-9]*`
-
-        """
-        hg_tags, _, ret = self.do_ex(
-            "hg log -r reverse(ancestors(.)) -T {tags}{if(tags, ' ', '')}"
-        )
-        if ret:
-            return None, None, None
-        hg_tags = hg_tags.split()
-
-        git_tags = {}
-        with open(os.path.join(self.path, ".hg/git-tags"), "r") as file:
-            for line in file:
-                node, tag = line.split()
-                git_tags[tag] = node
-
-        # find the first hg tag which is also a git tag
-        # TODO: also check for match *[0-9]*
-        for tag in hg_tags:
-            if tag in git_tags:
-                break
-
-        out, _, ret = self.do_ex("hg log -r .::" + tag + " -T .")
-        if ret:
-            return None, None, None
-        distance = len(out) - 1
-
-        node = self.node()
-        desc = f"{tag}-{distance}-g{node}"
-
-        if self.is_dirty():
-            desc += "-dirty"
-
-        return desc, None, 0
-
-
 def warn_on_shallow(wd):
     """experimental, may change at any time"""
     if wd.is_shallow():
@@ -225,6 +114,8 @@ def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
 
     wd = GitWorkdir.from_potential_worktree(config.absolute_root)
     if wd is None:
+        from .hg_git import GitWorkdirHgClient
+
         wd = GitWorkdirHgClient.from_potential_worktree(config.absolute_root)
     if wd is None:
         return

--- a/src/setuptools_scm/git.py
+++ b/src/setuptools_scm/git.py
@@ -137,11 +137,17 @@ class GitWorkdirHgClient(GitWorkdir):
         if hg_node is None:
             return
 
+        git_node = None
         with open(os.path.join(self.path, ".hg/git-mapfile"), "r") as file:
             for line in file:
                 if hg_node in line:
                     git_node, hg_node = line.split()
                     break
+
+        if git_node is None:
+            # trying again after hg -> git
+            self.do_ex("hg gexport")
+            return self.node()
 
         return git_node[:7]
 

--- a/src/setuptools_scm/git.py
+++ b/src/setuptools_scm/git.py
@@ -15,6 +15,7 @@ DEFAULT_DESCRIBE = "git describe --dirty --tags --long --match *[0-9]*"
 
 class GitWorkdir:
     """experimental, may change at any time"""
+
     COMMAND = "git"
 
     def __init__(self, path):
@@ -59,7 +60,7 @@ class GitWorkdir:
     def get_head_date(self):
         timestamp, err, ret = self.do_ex("git log -n 1 HEAD --format=%cI")
         if ret:
-            trace("branch err", timestamp, err, ret)
+            trace("timestamp err", timestamp, err, ret)
             return
         # TODO, when dropping python3.6 use fromiso
         date_part = timestamp.split("T")[0]
@@ -200,8 +201,7 @@ def fail_on_shallow(wd):
     """experimental, may change at any time"""
     if wd.is_shallow():
         raise ValueError(
-            "%r is shallow, please correct with "
-            '"git fetch --unshallow"' % wd.path
+            f'{wd.path} is shallow, please correct with "git fetch --unshallow"'
         )
 
 
@@ -211,7 +211,6 @@ def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
     """
     if not config:
         config = Configuration(root=root)
-
 
     wd = GitWorkdir.from_potential_worktree(config.absolute_root)
     if wd is None:
@@ -229,8 +228,6 @@ def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
     else:
         out, unused_err, ret = wd.default_describe()
 
-    node_date = wd.get_head_date() or date.today()
-
     if ret:
         # If 'git git_describe_command' failed, try to get the information otherwise.
         branch, branch_err, branch_ret = wd.do_ex("git symbolic-ref --short HEAD")
@@ -241,48 +238,33 @@ def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
         rev_node = wd.node()
         dirty = wd.is_dirty()
 
-        if rev_node is None:
-            return meta(
-                "0.0",
-                distance=0,
-                node_date=node_date,
-                dirty=dirty,
-                branch=branch,
-                config=config,
-            )
+        tag = "0.0"
 
-        return meta(
-            "0.0",
-            distance=wd.count_all_nodes(),
-            node="g" + rev_node,
-            dirty=dirty,
-            branch=wd.get_branch(),
-            node_date=node_date,
-            config=config,
-        )
+        if rev_node is None:
+            distance = 0
+        else:
+            distance = wd.count_all_nodes()
+            branch = wd.get_branch()
+            node = "g" + rev_node
+
     else:
-        tag, number, node, dirty = _git_parse_describe(out)
+        tag, distance, node, dirty = _git_parse_describe(out)
 
         branch = wd.get_branch()
-        if number:
-            return meta(
-                tag,
-                config=config,
-                distance=number,
-                node=node,
-                dirty=dirty,
-                branch=branch,
-                node_date=node_date,
-            )
-        else:
-            return meta(
-                tag,
-                config=config,
-                node=node,
-                node_date=node_date,
-                dirty=dirty,
-                branch=branch,
-            )
+        if distance == 0:
+            distance = None
+
+    node_date = wd.get_head_date() or date.today()
+
+    return meta(
+        tag,
+        branch=branch,
+        node=node,
+        node_date=node_date,
+        distance=distance,
+        dirty=dirty,
+        config=config,
+    )
 
 
 def _git_parse_describe(describe_output):

--- a/src/setuptools_scm/git.py
+++ b/src/setuptools_scm/git.py
@@ -46,7 +46,6 @@ class GitWorkdir(Workdir):
         branch, err, ret = self.do_ex("git rev-parse --abbrev-ref HEAD")
         if ret:
             trace("branch err", branch, err, ret)
-            # TODO: understand the diff between these 2 commands
             branch, err, ret = self.do_ex("git symbolic-ref --short HEAD")
             if ret:
                 trace("branch err (symbolic-ref)", branch, err, ret)

--- a/src/setuptools_scm/git.py
+++ b/src/setuptools_scm/git.py
@@ -15,8 +15,10 @@ DEFAULT_DESCRIBE = "git describe --dirty --tags --long --match *[0-9]*"
 
 class GitWorkdir:
     """experimental, may change at any time"""
+    COMMAND = "git"
 
     def __init__(self, path):
+        require_command(self.COMMAND)
         self.path = path
 
     def do_ex(self, cmd):
@@ -81,6 +83,105 @@ class GitWorkdir:
         revs, _, _ = self.do_ex("git rev-list HEAD")
         return revs.count("\n") + 1
 
+    def default_describe(self):
+        return self.do_ex(DEFAULT_DESCRIBE)
+
+
+class GitWorkdirHgClient(GitWorkdir):
+    COMMAND = "hg"
+
+    @classmethod
+    def from_potential_worktree(cls, wd):
+        root, _, ret = do_ex("hg root", wd)
+        if ret:
+            return
+        return cls(root)
+
+    def is_dirty(self):
+        out, _, _ = self.do_ex("hg id -T '{dirty}'")
+        return bool(out)
+
+    def get_branch(self):
+        branch, err, ret = self.do_ex("hg id -T {bookmarks}")
+        if ret:
+            trace("branch err", branch, err, ret)
+            return
+        return branch
+
+    def get_head_date(self):
+        date_part, err, ret = self.do_ex("hg log -r . -T {shortdate(date)}")
+        if ret:
+            trace("head date err", date_part, err, ret)
+            return
+        return datetime.strptime(date_part, r"%Y-%m-%d").date()
+
+    def is_shallow(self):
+        return False
+
+    def fetch_shallow(self):
+        pass
+
+    def get_hg_node(self):
+        rev_node, _, ret = self.do_ex("hg log -r . -T {node}")
+        if not ret:
+            return rev_node
+
+    def node(self):
+        hg_node = self.get_hg_node()
+        if hg_node is None:
+            return
+
+        with open(os.path.join(self.path, ".hg/git-mapfile"), "r") as file:
+            for line in file:
+                if hg_node in line:
+                    git_node, hg_node = line.split()
+                    break
+
+        return git_node[:7]
+
+    def count_all_nodes(self):
+        revs, _, _ = self.do_ex("hg log -r 'ancestors(.)' -T '.'")
+        return len(revs) + 1
+
+    def default_describe(self):
+        """
+        Tentative to reproduce the output of
+
+        `git describe --dirty --tags --long --match *[0-9]*`
+
+        """
+        hg_tags, _, ret = self.do_ex(
+            "hg log -r reverse(ancestors(.)) -T {tags}{if(tags, ' ', '')}"
+        )
+
+        if ret:
+            return None, None, None
+
+        git_tags = {}
+        with open(os.path.join(self.path, ".hg/git-tags"), "r") as file:
+            for line in file:
+                node, tag = line.split()
+                git_tags[tag] = node
+
+        # find the first hg tag which is also a git tag
+        # TODO: also check for match *[0-9]*
+        for tag in hg_tags:
+            if tag in git_tags:
+                break
+
+        out, _, ret = self.do_ex("hg log -r .:" + tag + " -T .")
+        if ret:
+            return None, None, None
+        distance = len(out) - 1
+
+        rev_node = self.node()
+        desc = f"{tag}-{distance}-g{rev_node}"
+
+        if self.is_dirty():
+            desc += "-dirty"
+
+        return desc, None, 0
+
 
 def warn_on_shallow(wd):
     """experimental, may change at any time"""
@@ -99,31 +200,35 @@ def fail_on_shallow(wd):
     """experimental, may change at any time"""
     if wd.is_shallow():
         raise ValueError(
-            "%r is shallow, please correct with " '"git fetch --unshallow"' % wd.path
+            "%r is shallow, please correct with "
+            '"git fetch --unshallow"' % wd.path
         )
 
 
-def parse(
-    root, describe_command=DEFAULT_DESCRIBE, pre_parse=warn_on_shallow, config=None
-):
+def parse(root, describe_command=None, pre_parse=warn_on_shallow, config=None):
     """
     :param pre_parse: experimental pre_parse action, may change at any time
     """
     if not config:
         config = Configuration(root=root)
 
-    require_command("git")
 
     wd = GitWorkdir.from_potential_worktree(config.absolute_root)
+    if wd is None:
+        wd = GitWorkdirHgClient.from_potential_worktree(config.absolute_root)
     if wd is None:
         return
     if pre_parse:
         pre_parse(wd)
 
-    if config.git_describe_command:
+    if config.git_describe_command is not None:
         describe_command = config.git_describe_command
 
-    out, unused_err, ret = wd.do_ex(describe_command)
+    if describe_command is not None:
+        out, unused_err, ret = wd.do_ex(describe_command)
+    else:
+        out, unused_err, ret = wd.default_describe()
+
     node_date = wd.get_head_date() or date.today()
 
     if ret:

--- a/src/setuptools_scm/hg.py
+++ b/src/setuptools_scm/hg.py
@@ -1,35 +1,118 @@
 import os
 from .config import Configuration
-from .utils import do, trace, data_from_mime, require_command
-from .version import meta, tags_to_versions
+from .utils import do_ex, do, trace, data_from_mime, require_command
+from .version import meta, tags_to_versions, tag_to_version
 
 
-def _hg_tagdist_normalize_tagcommit(config, tag, dist, node, branch):
-    dirty = node.endswith("+")
-    node = "h" + node.strip("+")
+class Workdir:
+    def __init__(self, path):
+        require_command(self.COMMAND)
+        self.path = path
 
-    # Detect changes since the specified tag
-    revset = (
-        "(branch(.)"  # look for revisions in this branch only
-        " and tag({tag!r})::."  # after the last tag
-        # ignore commits that only modify .hgtags and nothing else:
-        " and (merge() or file('re:^(?!\\.hgtags).*$'))"
-        " and not tag({tag!r}))"  # ignore the tagged commit itself
-    ).format(tag=tag)
-    if tag != "0.0":
-        commits = do(
-            ["hg", "log", "-r", revset, "--template", "{node|short}"],
-            config.absolute_root,
+    def do_ext(self, cmd):
+        return do_ex(cmd, cwd=self.path)
+
+    def do(self, cmd):
+        return do(cmd, cwd=self.path)
+
+
+class HgWorkdir(Workdir):
+
+    COMMAND = "hg"
+
+    @classmethod
+    def from_potential_worktree(cls, wd):
+        require_command(cls.COMMAND)
+        root, err, ret = do_ex("hg root", wd)
+        if ret:
+            print(err)
+            return
+        return cls(root)
+
+    def get_meta(self, config):
+
+        node, tags, bookmark, node_date = self.hg_log(
+            ".", "{node}\n{tag}\n{bookmark}\n{date|shortdate}"
+        ).split("\n")
+
+        branch, dirty, dirty_date = self.do(
+            ["hg", "id", "-T", "{branch}\n{if(dirty, 1, 0)}\n{date|shortdate}"]
+        ).split("\n")
+        dirty = bool(int(dirty))
+
+        if all(c == "0" for c in node):
+            trace("initial node", self.path)
+            return meta("0.0", config=config, dirty=dirty, branch=branch)
+
+        node = "h" + node[:7]
+
+        tags = tags.split()
+        if "tip" in tags:
+            # tip is not a real tag
+            tags = tags.remove("tip")
+
+        if tags:
+            tag = tags[0]
+            tag = tag_to_version(tag)
+            if tag:
+                return meta(tag, dirty=dirty, branch=branch, config=config)
+
+        try:
+            tag = self.get_latest_normalizable_tag()
+            dist = self.get_distance_revs(tag)
+            if tag == "null":
+                tag = "0.0"
+                dist = int(dist) + 1
+
+            if self.check_changes_since_tag(tag) or dirty:
+                return meta(
+                    tag,
+                    distance=dist,
+                    node=node,
+                    dirty=dirty,
+                    branch=branch,
+                    config=config,
+                )
+            else:
+                return meta(tag, config=config)
+
+        except ValueError:
+            pass  # unpacking failed, old hg
+
+    def hg_log(self, revset, template):
+        cmd = ["hg", "log", "-r", revset, "-T", template]
+        return self.do(cmd)
+
+    def get_latest_normalizable_tag(self):
+        # Gets all tags containing a '.' (see #229) from oldest to newest
+        outlines = self.hg_log(
+            revset="ancestors(.) and tag('re:\\.')",
+            template="{tags}{if(tags, '\n', '')}",
+        ).split()
+        if not outlines:
+            return "null"
+        tag = outlines[-1].split()[-1]
+        return tag
+
+    def get_distance_revs(self, rev1, rev2="."):
+        revset = f"({rev1}::{rev2})"
+        out = self.hg_log(revset, ".")
+        return len(out) - 1
+
+    def check_changes_since_tag(self, tag):
+
+        if tag == "0.0":
+            return True
+
+        revset = (
+            "(branch(.)"  # look for revisions in this branch only
+            f" and tag({tag!r})::."  # after the last tag
+            # ignore commits that only modify .hgtags and nothing else:
+            " and (merge() or file('re:^(?!\\.hgtags).*$'))"
+            f" and not tag({tag!r}))"  # ignore the tagged commit itself
         )
-    else:
-        commits = True
-    trace("normalize", locals())
-    if commits or dirty:
-        return meta(
-            tag, distance=dist, node=node, dirty=dirty, branch=branch, config=config
-        )
-    else:
-        return meta(tag, config=config)
+
+        return bool(self.hg_log(revset, "."))
 
 
 def parse(root, config=None):
@@ -41,56 +124,12 @@ def parse(root, config=None):
     if not config:
         config = Configuration(root=root)
 
-    require_command("hg")
-    identity_data = do("hg id -i -b -t", config.absolute_root).split()
-    if not identity_data:
+    wd = HgWorkdir.from_potential_worktree(config.absolute_root)
+
+    if wd is None:
         return
-    node = identity_data.pop(0)
-    branch = identity_data.pop(0)
-    if "tip" in identity_data:
-        # tip is not a real tag
-        identity_data.remove("tip")
-    tags = tags_to_versions(identity_data)
-    dirty = node[-1] == "+"
-    if tags:
-        return meta(tags[0], dirty=dirty, branch=branch, config=config)
 
-    if node.strip("+") == "0" * 12:
-        trace("initial node", config.absolute_root)
-        return meta("0.0", config=config, dirty=dirty, branch=branch)
-
-    try:
-        tag = get_latest_normalizable_tag(config.absolute_root)
-        dist = get_graph_distance(config.absolute_root, tag)
-        if tag == "null":
-            tag = "0.0"
-            dist = int(dist) + 1
-        return _hg_tagdist_normalize_tagcommit(config, tag, dist, node, branch)
-    except ValueError:
-        pass  # unpacking failed, old hg
-
-
-def get_latest_normalizable_tag(root):
-    # Gets all tags containing a '.' (see #229) from oldest to newest
-    cmd = [
-        "hg",
-        "log",
-        "-r",
-        "ancestors(.) and tag('re:\\.')",
-        "--template",
-        "{tags}\n",
-    ]
-    outlines = do(cmd, root).split()
-    if not outlines:
-        return "null"
-    tag = outlines[-1].split()[-1]
-    return tag
-
-
-def get_graph_distance(root, rev1, rev2="."):
-    cmd = ["hg", "log", "-q", "-r", f"{rev1}::{rev2}"]
-    out = do(cmd, root)
-    return len(out.strip().splitlines()) - 1
+    return wd.get_meta(config)
 
 
 def archival_to_version(data, config=None):

--- a/src/setuptools_scm/hg.py
+++ b/src/setuptools_scm/hg.py
@@ -24,10 +24,19 @@ class HgWorkdir(Workdir):
             ".", "{node}\n{tag}\n{bookmark}\n{date|shortdate}"
         ).split("\n")
 
+        # TODO: support bookmarks and topics (but nowadays bookmarks are
+        # mainly used to emulate Git branches, which is already supported with
+        # the dedicated class GitWorkdirHgClient)
+
         branch, dirty, dirty_date = self.do(
             ["hg", "id", "-T", "{branch}\n{if(dirty, 1, 0)}\n{date|shortdate}"]
         ).split("\n")
         dirty = bool(int(dirty))
+
+        if dirty:
+            date = dirty_date
+        else:
+            date = node_date
 
         if all(c == "0" for c in node):
             trace("initial node", self.path)
@@ -61,6 +70,7 @@ class HgWorkdir(Workdir):
                     dirty=dirty,
                     branch=branch,
                     config=config,
+                    node_date=date
                 )
             else:
                 return meta(tag, config=config)

--- a/src/setuptools_scm/hg.py
+++ b/src/setuptools_scm/hg.py
@@ -33,6 +33,11 @@ def _hg_tagdist_normalize_tagcommit(config, tag, dist, node, branch):
 
 
 def parse(root, config=None):
+    if os.path.exists(os.path.join(root, ".hg/git")):
+        from .git import parse as git_parse
+
+        return git_parse(root, config=config)
+
     if not config:
         config = Configuration(root=root)
 

--- a/src/setuptools_scm/hg.py
+++ b/src/setuptools_scm/hg.py
@@ -70,7 +70,7 @@ class HgWorkdir(Workdir):
                     dirty=dirty,
                     branch=branch,
                     config=config,
-                    node_date=date
+                    node_date=date,
                 )
             else:
                 return meta(tag, config=config)

--- a/src/setuptools_scm/hg.py
+++ b/src/setuptools_scm/hg.py
@@ -1,4 +1,6 @@
 import os
+from pathlib import Path
+
 from .config import Configuration
 from .utils import do_ex, trace, data_from_mime, require_command
 from .version import meta, tag_to_version
@@ -116,9 +118,15 @@ class HgWorkdir(Workdir):
 
 def parse(root, config=None):
     if os.path.exists(os.path.join(root, ".hg/git")):
-        from .git import parse as git_parse
+        paths, _, ret = do_ex("hg path", root)
+        if not ret:
+            for line in paths.split("\n"):
+                if line.startswith("default ="):
+                    path = Path(line.split()[2])
+                    if path.name.endswith(".git") or (path / ".git").exists():
+                        from .git import parse as git_parse
 
-        return git_parse(root, config=config)
+                        return git_parse(root, config=config)
 
     if not config:
         config = Configuration(root=root)

--- a/src/setuptools_scm/hg.py
+++ b/src/setuptools_scm/hg.py
@@ -1,19 +1,8 @@
 import os
 from .config import Configuration
-from .utils import do_ex, do, trace, data_from_mime, require_command
-from .version import meta, tags_to_versions, tag_to_version
-
-
-class Workdir:
-    def __init__(self, path):
-        require_command(self.COMMAND)
-        self.path = path
-
-    def do_ext(self, cmd):
-        return do_ex(cmd, cwd=self.path)
-
-    def do(self, cmd):
-        return do(cmd, cwd=self.path)
+from .utils import do_ex, trace, data_from_mime, require_command
+from .version import meta, tag_to_version
+from .scm_workdir import Workdir
 
 
 class HgWorkdir(Workdir):

--- a/src/setuptools_scm/hg_git.py
+++ b/src/setuptools_scm/hg_git.py
@@ -1,0 +1,110 @@
+import os
+from datetime import datetime
+
+from .utils import trace, require_command, do_ex
+from .git import GitWorkdir
+from .hg import HgWorkdir
+
+
+class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
+    COMMAND = "hg"
+
+    @classmethod
+    def from_potential_worktree(cls, wd):
+        require_command(cls.COMMAND)
+        root, err, ret = do_ex("hg root", wd)
+        if ret:
+            print(err)
+            return
+        return cls(root)
+
+    def is_dirty(self):
+        out, _, _ = self.do_ex("hg id -T '{dirty}'")
+        return bool(out)
+
+    def get_branch(self):
+        branch, err, ret = self.do_ex("hg id -T {bookmarks}")
+        if ret:
+            trace("branch err", branch, err, ret)
+            return
+        return branch
+
+    def get_head_date(self):
+        date_part, err, ret = self.do_ex("hg log -r . -T {shortdate(date)}")
+        if ret:
+            trace("head date err", date_part, err, ret)
+            return
+        return datetime.strptime(date_part, r"%Y-%m-%d").date()
+
+    def is_shallow(self):
+        return False
+
+    def fetch_shallow(self):
+        pass
+
+    def get_hg_node(self):
+        node, _, ret = self.do_ex("hg log -r . -T {node}")
+        if not ret:
+            return node
+
+    def node(self):
+        hg_node = self.get_hg_node()
+        if hg_node is None:
+            return
+
+        git_node = None
+        with open(os.path.join(self.path, ".hg/git-mapfile"), "r") as file:
+            for line in file:
+                if hg_node in line:
+                    git_node, hg_node = line.split()
+                    break
+
+        if git_node is None:
+            # trying again after hg -> git
+            self.do_ex("hg gexport")
+            return self.node()
+
+        return git_node[:7]
+
+    def count_all_nodes(self):
+        revs, _, _ = self.do_ex("hg log -r 'ancestors(.)' -T '.'")
+        return len(revs) + 1
+
+    def default_describe(self):
+        """
+        Tentative to reproduce the output of
+
+        `git describe --dirty --tags --long --match *[0-9]*`
+
+        """
+        hg_tags, _, ret = self.do_ex(
+            "hg log -r reverse(ancestors(.)) -T {tags}{if(tags, ' ', '')}"
+        )
+        if ret:
+            return None, None, None
+        hg_tags = hg_tags.split()
+
+        git_tags = {}
+        with open(os.path.join(self.path, ".hg/git-tags"), "r") as file:
+            for line in file:
+                node, tag = line.split()
+                git_tags[tag] = node
+
+        # find the first hg tag which is also a git tag
+        # TODO: also check for match *[0-9]*
+        for tag in hg_tags:
+            if tag in git_tags:
+                break
+
+        out, _, ret = self.do_ex("hg log -r " + tag + "::. -T .")
+        if ret:
+            return None, None, None
+        distance = len(out) - 1
+
+        node = self.node()
+        desc = f"{tag}-{distance}-g{node}"
+
+        if self.is_dirty():
+            desc += "-dirty"
+
+        return desc, None, 0

--- a/src/setuptools_scm/hg_git.py
+++ b/src/setuptools_scm/hg_git.py
@@ -78,7 +78,14 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
 
         """
         hg_tags, _, ret = self.do_ex(
-            "hg log -r reverse(ancestors(.)) -T {tags}{if(tags, ' ', '')}"
+            [
+                "hg",
+                "log",
+                "-r",
+                "(reverse(ancestors(.)) and tag(r're:[0-9]'))",
+                "-T",
+                "{tags}{if(tags, ' ', '')}",
+            ]
         )
         if ret:
             return None, None, None
@@ -91,7 +98,6 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
                 git_tags[tag] = node
 
         # find the first hg tag which is also a git tag
-        # TODO: also check for match *[0-9]*
         for tag in hg_tags:
             if tag in git_tags:
                 break

--- a/src/setuptools_scm/hg_git.py
+++ b/src/setuptools_scm/hg_git.py
@@ -47,22 +47,30 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
         if not ret:
             return node
 
-    def node(self):
-        hg_node = self.get_hg_node()
-        if hg_node is None:
-            return
-
+    def _hg2git(self, hg_node):
         git_node = None
         with open(os.path.join(self.path, ".hg/git-mapfile")) as file:
             for line in file:
                 if hg_node in line:
                     git_node, hg_node = line.split()
                     break
+        return git_node
+
+    def node(self):
+        hg_node = self.get_hg_node()
+        if hg_node is None:
+            return
+
+        git_node = self._hg2git(hg_node)
 
         if git_node is None:
             # trying again after hg -> git
             self.do_ex("hg gexport")
-            return self.node()
+            git_node = self._hg2git(hg_node)
+
+            if git_node is None:
+                trace("Cannot get git node so we use hg node", hg_node)
+                return hg_node
 
         return git_node[:7]
 
@@ -90,6 +98,9 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
         if ret:
             return None, None, None
         hg_tags = hg_tags.split()
+
+        if not hg_tags:
+            return None, None, None
 
         git_tags = {}
         with open(os.path.join(self.path, ".hg/git-tags")) as file:

--- a/src/setuptools_scm/hg_git.py
+++ b/src/setuptools_scm/hg_git.py
@@ -70,13 +70,18 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
 
             if git_node is None:
                 trace("Cannot get git node so we use hg node", hg_node)
+
+                if hg_node == "0" * len(hg_node):
+                    # mimick Git behavior
+                    return None
+
                 return hg_node
 
         return git_node[:7]
 
     def count_all_nodes(self):
         revs, _, _ = self.do_ex("hg log -r 'ancestors(.)' -T '.'")
-        return len(revs) + 1
+        return len(revs)
 
     def default_describe(self):
         """
@@ -113,7 +118,7 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
             if tag in git_tags:
                 break
 
-        out, _, ret = self.do_ex("hg log -r " + tag + "::. -T .")
+        out, _, ret = self.do_ex(["hg", "log", "-r", f"'{tag}'::.", "-T", "."])
         if ret:
             return None, None, None
         distance = len(out) - 1

--- a/src/setuptools_scm/hg_git.py
+++ b/src/setuptools_scm/hg_git.py
@@ -53,7 +53,7 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
             return
 
         git_node = None
-        with open(os.path.join(self.path, ".hg/git-mapfile"), "r") as file:
+        with open(os.path.join(self.path, ".hg/git-mapfile")) as file:
             for line in file:
                 if hg_node in line:
                     git_node, hg_node = line.split()
@@ -92,7 +92,7 @@ class GitWorkdirHgClient(GitWorkdir, HgWorkdir):
         hg_tags = hg_tags.split()
 
         git_tags = {}
-        with open(os.path.join(self.path, ".hg/git-tags"), "r") as file:
+        with open(os.path.join(self.path, ".hg/git-tags")) as file:
             for line in file:
                 node, tag = line.split()
                 git_tags[tag] = node

--- a/src/setuptools_scm/scm_workdir.py
+++ b/src/setuptools_scm/scm_workdir.py
@@ -1,0 +1,13 @@
+from .utils import do_ex, do, require_command
+
+
+class Workdir:
+    def __init__(self, path):
+        require_command(self.COMMAND)
+        self.path = path
+
+    def do_ex(self, cmd):
+        return do_ex(cmd, cwd=self.path)
+
+    def do(self, cmd):
+        return do(cmd, cwd=self.path)

--- a/testing/conftest.py
+++ b/testing/conftest.py
@@ -95,3 +95,29 @@ def wd(tmp_path):
     target_wd = tmp_path.resolve() / "wd"
     target_wd.mkdir()
     return Wd(target_wd)
+
+
+@pytest.fixture
+def repositories_hg_git(tmp_path):
+    from setuptools_scm.utils import do
+
+    tmp_path = tmp_path.resolve()
+    path_git = tmp_path / "repo_git"
+    path_git.mkdir()
+
+    wd = Wd(path_git)
+    wd("git init")
+    wd("git config user.email test@example.com")
+    wd('git config user.name "a test"')
+    wd.add_command = "git add ."
+    wd.commit_command = "git commit -m test-{reason}"
+
+    path_hg = tmp_path / "repo_hg"
+    do(f"hg clone {path_git} {path_hg}")
+    assert path_hg.exists()
+
+    wd_hg = Wd(path_hg)
+    wd_hg.add_command = "hg add ."
+    wd_hg.commit_command = 'hg commit -m test-{reason} -u test -d "0 0"'
+
+    return wd_hg, wd

--- a/testing/conftest.py
+++ b/testing/conftest.py
@@ -113,8 +113,11 @@ def repositories_hg_git(tmp_path):
     wd.commit_command = "git commit -m test-{reason}"
 
     path_hg = tmp_path / "repo_hg"
-    do(f"hg clone {path_git} {path_hg}")
+    do(f"hg clone {path_git} {path_hg} --config extensions.hggit=")
     assert path_hg.exists()
+
+    with open(path_hg / ".hg/hgrc", "a") as file:
+        file.write("[extensions]\nhggit =\n")
 
     wd_hg = Wd(path_hg)
     wd_hg.add_command = "hg add ."

--- a/testing/test_hg_git.py
+++ b/testing/test_hg_git.py
@@ -1,6 +1,18 @@
 import pytest
+from setuptools_scm.utils import has_command, do_ex
+
+python_hg, err, ret = do_ex("hg debuginstall --template {pythonexe}")
+
+if ret:
+    skip_no_hggit = True
+else:
+    out, err, ret = do_ex([python_hg.strip(), "-c", "import hggit"])
+    print(out, err, ret)
+    skip_no_hggit = bool(ret)
 
 
+@pytest.mark.skipif(not has_command("hg", warn=False), reason="hg executable not found")
+@pytest.mark.skipif(skip_no_hggit, reason="hg-git not installed")
 def test_base(repositories_hg_git):
     wd, wd_git = repositories_hg_git
 

--- a/testing/test_hg_git.py
+++ b/testing/test_hg_git.py
@@ -1,0 +1,60 @@
+import pytest
+
+
+def test_base(repositories_hg_git):
+    wd, wd_git = repositories_hg_git
+
+    assert wd_git.version == "0.1.dev0"
+    assert wd.version == "0.1.dev0"
+
+    wd_git.commit_testfile()
+    wd("hg pull -u")
+
+    version_git = wd_git.version
+    version = wd.version
+
+    assert version_git.startswith("0.1.dev1+g")
+    assert version.startswith("0.1.dev1+g")
+
+    assert not version_git.endswith("1-")
+    assert not version.endswith("1-")
+
+    wd_git("git tag v0.1")
+    wd("hg pull -u")
+    assert wd_git.version == "0.1"
+    assert wd.version == "0.1"
+
+    wd_git.write("test.txt", "test2")
+    wd.write("test.txt", "test2")
+    assert wd_git.version.startswith("0.2.dev0+g")
+    assert wd.version.startswith("0.2.dev0+g")
+
+    wd_git.commit_testfile()
+    wd("hg pull")
+    wd("hg up -C")
+    assert wd_git.version.startswith("0.2.dev1+g")
+    assert wd.version.startswith("0.2.dev1+g")
+
+    wd_git("git tag version-0.2")
+    wd("hg pull -u")
+    assert wd_git.version.startswith("0.2")
+    assert wd.version.startswith("0.2")
+
+    wd_git.commit_testfile()
+    wd_git("git tag version-0.2.post210+gbe48adfpost3+g0cc25f2")
+    wd("hg pull -u")
+    with pytest.warns(
+        UserWarning, match="tag '.*' will be stripped of its suffix '.*'"
+    ):
+        assert wd_git.version.startswith("0.2")
+
+    with pytest.warns(
+        UserWarning, match="tag '.*' will be stripped of its suffix '.*'"
+    ):
+        assert wd.version.startswith("0.2")
+
+    wd_git.commit_testfile()
+    wd_git("git tag 17.33.0-rc")
+    wd("hg pull -u")
+    assert wd_git.version == "17.33.0rc0"
+    assert wd.version == "17.33.0rc0"

